TypeCobol.Analysis/Graph/ControlFlowGraph.cs

```diff
@@ -0,0 +1,170 @@
+﻿using System;
+using System.Collections.Generic;
+
+namespace TypeCobol.Analysis.Graph
+{
+    /// <summary>
+    /// A set of basic blocks, each of which has a list of successor blocks and some other information.
+    /// Each block consists of a list of instructions, each of which can point to previous instructions that compute the operands it consumes.
+    /// </summary>
+    /// <typeparam name="N"></typeparam>
+    /// <typeparam name="D"></typeparam>
+    public class ControlFlowGraph<N, D>
+    {
+        /// <summary>
+        /// BasicBlock callback type.
+        /// </summary>
+        /// <param name="block">The BasicBlock</param>
+        /// <param name="cfg">The Control Flow Graph in which the basic Block belongs to.</param>
+        /// <returns>true if ok, false otherwise</returns>
+        public delegate bool BasicBlockCallback(BasicBlock<N, D> block, ControlFlowGraph<N, D> cfg);
+
+        /// <summary>
+        /// Flag on a Cfg.
+        /// </summary>
+        [Flags]
+        public enum Flags : uint
+        {
+            Compound = 0x01 << 0,       //Flag to indicate that this Cfg graph has subgraphs.
+        }
+
+        /// <summary>
+        /// Symbol Flags.
+        /// </summary>
+        public Flags Flag
+        {
+            get;
+            internal set;
+        }
+
+        /// <summary>
+        /// Set a set of flags to true or false.
+        /// </summary>
+        /// <param name="flag"></param>
+        /// <param name="value"></param>
+        internal virtual void SetFlag(Flags flag, bool value)
+        {
+            this.Flag = value ? (Flags)(this.Flag | flag)
+                              : (Flags)(this.Flag & ~flag);
+        }
+
+        /// <summary>
+        /// Determines if the given flag is set.
+        /// </summary>
+        /// <param name="flag">The flag to be tested</param>
+        /// <returns>true if yes, false otherwise.</returns>
+        public bool HasFlag(Flags flag)
+        {
+            return (this.Flag & flag) != 0;
+        }
+
+        /// <summary>
+        /// Root blocks. Usually it has a single item which is the first block in the program.
+        /// But Exception handlers are also stored as root blocks.
+        /// </summary>
+        public List<BasicBlock<N, D>> RootBlocks
+        {
+            get;
+            internal set;
+        }
+
+        /// <summary>
+        /// All blocks. A list of basic blocks in the graph.
+        /// </summary>
+        public List<BasicBlock<N, D>> AllBlocks
+        {
+            get;
+            internal set;
+        }
+
+        /// <summary>
+        /// A map from Node to corresponding basic block.
+        /// </summary>
+        public Dictionary<N, BasicBlock<N,D>> BlockFor
```

Comment:
This dictionary has its values set but never read. It can be used in the various `ResolvePending*` methods to get the block back from a node, this would allow us to store only nodes to be resolved instead of Tuples.

Reply:
Hum.. block for node is usefull for instance for:

https://github.com/TypeCobolTeam/TypeCobol/pull/1493
https://github.com/TypeCobolTeam/TypeCobol/issues/1449